### ISSUE TITLE
bug: select, long text does not truncate

### ISSUE THREAD

**Issue by matthew2564** [ionic-team/ionic-framework#27081]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already report this problem, without success.

### Ionic Framework Version

v7.x

### Current Behavior

Ionic now suggests adding a `label` or `aria-label` to tags such as `ion-select` which is highlighted via this console warning

```
[Ionic Warning]: ion-select now requires providing a label with either the "label" property or the "aria-label" attribute. To migrate, remove any usage of "ion-label" and pass the label text to either the "label" property or the "aria-label" attribute.

Example: <ion-select label="Favorite Color">...</ion-select>
Example with aria-label: <ion-select aria-label="Favorite Color">...</ion-select>
```

Upon adding an `aria-label` attribute to the `ion-select`, the positioning of the up / down carets on the right hand side comes inwards. Here is a basic code example

```
<ion-header [translucent]="true">
  <ion-toolbar>
    <ion-title>
      Blank
    </ion-title>
  </ion-toolbar>
</ion-header>

<ion-content [fullscreen]="true">
    <ion-row style="border: 3px solid black;">

      <ion-col class="ion-align-self-center" style="border: 3px solid red;" size="4">
        <label>Fruit</label>
      </ion-col>

      <ion-col>
        <ion-select placeholder="Select a fruit" fill="outline" aria-label="Fruit" style="border: 3px solid blue;">
          <ion-select-option value="apple">Apple</ion-select-option>
          <ion-select-option value="banana">Banana</ion-select-option>
          <ion-select-option value="orange">Orange</ion-select-option>
        </ion-select>
      </ion-col>

    </ion-row>
</ion-content>
```

The result of that being...

<img width="604" alt="Screenshot 2023-04-01 at 15 58 18" src="https://user-images.githubusercontent.com/33056264/229297134-a7115cbb-64dd-436d-99fa-2c0f2154dcaf.png">

### Expected Behavior

I would expect the position of the carets to remain the same as without the `aria-label` attribute. 

In the below example, the carets remain in the position expected.

```
<ion-header [translucent]="true">
  <ion-toolbar>
    <ion-title>
      Blank
    </ion-title>
  </ion-toolbar>
</ion-header>

<ion-content [fullscreen]="true">
    <ion-row style="border: 3px solid black;">

      <ion-col class="ion-align-self-center" style="border: 3px solid red;" size="4">
        <label id="fruit-label">Fruit</label>
      </ion-col>

      <ion-col>
        <ion-select placeholder="Select a fruit" fill="outline" style="border: 3px solid blue;">
          <ion-select-option value="apple">Apple</ion-select-option>
          <ion-select-option value="banana">Banana</ion-select-option>
          <ion-select-option value="orange">Orange</ion-select-option>
        </ion-select>
      </ion-col>

    </ion-row>
</ion-content>
```
The result of that being...

<img width="594" alt="Screenshot 2023-04-01 at 15 58 27" src="https://user-images.githubusercontent.com/33056264/229297306-35c18483-69cc-4e3e-87f0-fe80ec9438a4.png">

### Steps to Reproduce

1. Add snippet provided above in the `Current Behaviour` section

### Ionic Info

```
Ionic:

   Ionic CLI                     : 6.20.8 (/opt/homebrew/lib/node_modules/@ionic/cli)
   Ionic Framework               : not installed
   @angular-devkit/build-angular : 15.2.4
   @angular-devkit/schematics    : 15.2.4
   @angular/cli                  : 15.2.4
   @ionic/angular-toolkit        : 9.0.0

Capacitor:

   Capacitor CLI      : 4.7.3
   @capacitor/android : not installed
   @capacitor/core    : 4.7.3
   @capacitor/ios     : not installed

Utility:

   cordova-res : not installed globally
   native-run  : 1.7.2

System:

   NodeJS : v18.10.0 (/Users/matt/.n/bin/node)
   npm    : 8.19.2
   OS     : macOS Unknown
```

**Comment by judsonmusic** [comment]
I was struggling with this today. I had to do a workaround by setting ion-item and ion-select to 100% width

**Comment by matthew2564** [comment]
Thanks for the workaround @judsonmusic, however there are further issues off the back of this. 

```
<ion-content [fullscreen]="true">
    <ion-row style="border: 3px solid black;">
      <ion-col class="ion-align-self-center" style="border: 3px solid red;" size="4">
        <label id="fruit-label">Fruit</label>
      </ion-col>

      <ion-col size="3">
        <ion-select
          placeholder="Select a fruit"
          aria-label="fruit"
          labelPlacement="stacked"
          multiple="true"
          fill="outline"
          style="border: 3px solid blue;"
        >
          <ion-select-option value="apple">Apple</ion-select-option>
          <ion-select-option value="banana">Banana</ion-select-option>
          <ion-select-option value="orange">Orange</ion-select-option>
        </ion-select>
      </ion-col>
    </ion-row>
</ion-content>
```
This achieves the desired result mentioned above with the output of
<img width="598" alt="Screenshot 2023-04-02 at 10 01 56" src="https://user-images.githubusercontent.com/33056264/229343251-cbb17f03-17ad-47bc-9580-890713abc43b.png">


However, if the length of the selected value exceeds that of the `ion-select` area, there is no longer an ellipsis / ... to truncate the value, this is also a problem when you have multiple selectable options. See below (following code snippet is the same as the one above).

```
<ion-content [fullscreen]="true">
    <ion-row style="border: 3px solid black;">
      <ion-col class="ion-align-self-center" style="border: 3px solid red;" size="4">
        <label id="fruit-label">Fruit</label>
      </ion-col>

      <ion-col size="3">
        <ion-select
          placeholder="Select a fruit"
          aria-label="fruit"
          labelPlacement="stacked"
          multiple="true"
          fill="outline"
          style="border: 3px solid blue;"
        >
          <ion-select-option value="apple">Apple</ion-select-option>
          <ion-select-option value="banana">Banana</ion-select-option>
          <ion-select-option value="orange">Orange</ion-select-option>
        </ion-select>
      </ion-col>
    </ion-row>
</ion-content>
```
When selecting all 3 items, this is how it is rendered (where the carets overlay the selected values)
<img width="595" alt="Screenshot 2023-04-02 at 10 01 50" src="https://user-images.githubusercontent.com/33056264/229343398-4a98cfe2-c802-4567-b526-e19282784734.png">

With only the removal of the `aria-label` attribute, the ellipsis is restored like so
<img width="602" alt="Screenshot 2023-04-02 at 10 02 21" src="https://user-images.githubusercontent.com/33056264/229343447-5e825421-1b5c-4a2c-845f-f992b4eeb42e.png">

In regards to the `ion-select` itself, it is frustrating that the `aria-labelledby` does not suppress the warning. As a result, I'm now having to re-label each `ion-select` instead of pointing it at the label associated.

**Comment by liamdebeasi** [comment]
Thanks for the report. Adding `aria-label` or `label` opts in to the new form control syntax which has a different architecture under the hood. The big difference is `ion-select` contains both the label and the form control. So in this case, you should be using the `label` property on `ion-select` instead of using a separate `<label>` and an `aria-label`.

Here is an example: https://codepen.io/liamdebeasi/pen/yLxmqdR

You can control the order of the label/control using `labelPlacement`. You can also control how the label and control are packed on a line by using the `justify` property.

**Comment by judsonmusic** [comment]
@liamdebeasi can you provide an example using the justify property?

Also, my company uses position stacked for everything so it might be worthwhile making sure the docs are updated for this as well. I created a ticket based on the fact that the stacked label for the ion select is off by a few pixels. Just FYI compared to the ion input.

**Comment by judsonmusic** [comment]
That’s correct. It looks like two bugs that I posted actually are going to be fixed. Can you let me know when they will be released?JudsonOn Apr 3, 2023, at 3:04 PM, Liam DeBeasi ***@***.***> wrote:﻿
Sure, we have some playgrounds that show how the justify property works: https://ionicframework.com/docs/api/select#justification

Also, my company uses position stacked for everything so it might be worthwhile making sure the docs are updated for this as well. I created a ticket based on the fact that the stacked label for the ion select is off by a few pixels. Just FYI compared to the ion input.

This is #27086, correct?

**Comment by liamdebeasi** [comment]
Those issues have been labeled as bugs, but they have not been fixed yet. However, once the fixes merge we will post on each thread.

**Comment by judsonmusic** [comment]
@liamdebeasi yes, that is correct. I have two other bugs file that are very similar and it looks like they were indeed marked as bugs. At what point will these be released or how will I know so I can pull latest on ionic? Also, I don't know if you'd be willing to do it but I have been doing ionic since it came out. I'm an architect. I was wondering if it's possible to screen share and show you a couple of things before I actually post potential bugs for them. Let me know if that's a possibility you can email me at Judsonmusic@me.com I am on LinkedIn under Judson Terrell.

**Comment by liamdebeasi** [comment]
We don't provide public timelines on bug fixes since those timelines sometimes change. However, once a bug fix has been merged into `main` it is usually (but not always) released in the next patch release. We do patch releases once a week.

I'm fairly active on Discord if you wanted to ask your questions there: https://discord.com/invite/UPYYRhtyzp. Otherwise, feel free to open separate issues here on GitHub.

**Comment by judsonmusic** [comment]
@liamdebeasi I joined the discord. What is your discord username? I was going to ping you minors under Judson all capital letters. If you want to add me I'll ping you when I'm in front of my desk in a little bit.

**Comment by matthew2564** [comment]
> Thanks for the report. Adding `aria-label` or `label` opts in to the new form control syntax which has a different architecture under the hood. The big difference is `ion-select` contains both the label and the form control. So in this case, you should be using the `label` property on `ion-select` instead of using a separate `<label>` and an `aria-label`.
> 
> 
> 
> Here is an example: https://codepen.io/liamdebeasi/pen/yLxmqdR
> 
> 
> 
> You can control the order of the label/control using `labelPlacement`. You can also control how the label and control are packed on a line by using the `justify` property.

Does this not invalidate any custom HTML composition? 
The reason we are using label in a different context to the select/input itself is due to positioning on a page. i.e. ion-col's spacing elements apart. 

If we do not have true flexibility between separating the label to the element then it feels like this shouldn’t be a warning and more of an error.  

On another note, why does ‘aria-label’ have any impact on the positioning of the carets themselves? That query is more from inquisitive perspectiveto try and understand the underlying logic/tech more. 

Another point mentioned above is re the 'aria-labelledby', why is this not sufficient to suppress the warning produced via the element when using 'ion-select'?

**Comment by liamdebeasi** [comment]
If we added a way to customize the CSS of the label inside of `ion-select` when using the `label` property, would that resolve the issue for you?

**Comment by judsonmusic** [comment]
> If we added a way to customize the CSS of the label inside of `ion-select` when using the `label` property, would that resolve the issue for you?

@liamdebeasi I have always thought you should give devs a little more customization by adding a shadowClass attribute that would pierce the shadow dom element if someone truly wants to customize labels etc. For example we like to use a red asteriks to put next to required fields instead of blowing up the screen with red everywhere right out of the gate.

<ion-input shadowClass='myshadowDOMpiercingClass"></ion-input> or on a select etc. which would make it easier to do these customizations. The out of the box ionic solution tends to make people go "well that website just looks like everything else when bootstrap was popular" I dont want all of my ionic apps to have to look the same. Branding is important to UX designers and they dont like to be told, "we cannot do that because we are using ionic"

**Comment by judsonmusic** [comment]
@liamdebeasi i send you an FR on discord. If you could accept Id like to show you something

**Comment by muuvmuuv** [comment]
The issue we have is that it is nearly impossible rn to accomplish something like this. A part of that issue is that `native-wrapper` cannot be overridden.

![image](https://user-images.githubusercontent.com/11534760/229744357-e64333a7-a346-4224-bc2b-534d5214d8b3.png)

**Comment by matthew2564** [comment]
> If we added a way to customize the CSS of the label inside of `ion-select` when using the `label` property, would that resolve the issue for you?

Probably not, although I can imagine this type of customisation/control useful for others.

 The main issue I have is the lack of ellipsis meaning values inside the ion-select are overlaying the carets and not truncating. 

As mentioned above, without the `label`/`aria-label` attribute the ellipsis works as I would expect.

**Comment by judsonmusic** [comment]
> There are quite a few requests here, so I'd just like to make sure I'm focusing on the original request:
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/2721089/229836372-25ad784d-7f73-42ba-9306-e7014deec85d.png)
> 
> 
> 
> The original issue here is that the selected text overlaps the caret icon correct?

I definitely see a need for being able to customize the label. One of the biggest issues is that we were able to use label stacked, which put the label on top of the input field or select field and we are no longer able to do that so I had to re-factor our entire application and explain to people that we cannot do that anymore and they did not like it at all. I feel like the floating label should have a fixed option so it does not animate and stacked should do what it is supposed to do and stack over the top of the input as far as the original question you should make it to where the innertext compensate for the width of the label since the label encapsulates everything now so if someone wants it to be 100% with of an ion column which we do on a regular basis, the inner text of the label, or the inner native element compensates for the icon enough and defaults to create enough space for ellipses. @liamdebeasi I added you on Discord yesterday and would like to be able to show you a real world. Example of what I'm talking about to give you something to think about how you would solve both problems.
Thanks. You should see a request for me if you could just ping me back..

**Comment by liamdebeasi** [comment]
> > There are quite a few requests here, so I'd just like to make sure I'm focusing on the original request:
> > ![image](https://user-images.githubusercontent.com/2721089/229836372-25ad784d-7f73-42ba-9306-e7014deec85d.png)
> > The original issue here is that the selected text overlaps the caret icon correct?
> 
> I definitely see a need for being able to customize the label. One of the biggest issues is that we were able to use label stacked, which put the label on top of the input field or select field and we are no longer able to do that so I had to re-factor our entire application and explain to people that we cannot do that anymore and they did not like it at all. I feel like the floating label should have a fixed option so it does not animate and stacked should do what it is supposed to do and stack over the top of the input as far as the original question you should make it to where the innertext compensate for the width of the label since the label encapsulates everything now so if someone wants it to be 100% with of an ion column which we do on a regular basis, the inner text of the label, or the inner native element compensates for the icon enough and defaults to create enough space for ellipses. @liamdebeasi I added you on Discord yesterday and would like to be able to show you a real world. Example of what I'm talking about to give you something to think about how you would solve both problems. Thanks. You should see a request for me if you could just ping me back..


Could you file a separate issue for this? Having threads with multiple requests make it very hard to triage. I'd like to keep this thread focused on the issue I noted in https://github.com/ionic-team/ionic-framework/issues/27081#issuecomment-1496143351.

Re: Discord, if you post in the #ionic-general channel I can take a look.

**Comment by judsonmusic** [comment]
@liamdebeasi I was kind of hoping to screen share for a few minutes to show the behavior before I post anything useful or not useful, potentially, if you'd be willing to do that. I kind a need to verify that this is even a thing or that it's even possible I've been working with the ionic since it came out and seven has definitely caused some changes that were less than desired.

**Comment by liamdebeasi** [comment]
Filing an issue would be the best way to get any bugs addressed. If you're unsure if something is a bug feel free to file one anyways and we can take a look. I don't like discussing issues privately on Discord since the rest of the team has no visibility into the issues.

**Comment by matthew2564** [comment]
> There are quite a few requests here, so I'd just like to make sure I'm focusing on the original request:
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/2721089/229836372-25ad784d-7f73-42ba-9306-e7014deec85d.png)
> 
> 
> 
> The original issue here is that the selected text overlaps the caret icon correct?

Yes, that is correct 👍🏻

**Comment by liamdebeasi** [comment]
> > There are quite a few requests here, so I'd just like to make sure I'm focusing on the original request:
> > ![image](https://user-images.githubusercontent.com/2721089/229836372-25ad784d-7f73-42ba-9306-e7014deec85d.png)
> > The original issue here is that the selected text overlaps the caret icon correct?
> 
> Yes, that is correct 👍🏻

Thanks for confirming! Do you have more information on how I can reproduce this? I tried testing this on my end, but the text truncates with ellipses correctly: https://codepen.io/liamdebeasi/pen/gOBYZRR

**Comment by liamdebeasi** [comment]
What version of Chrome are you using?

**Comment by liamdebeasi** [comment]
Are you able to reproduce this issue in an Ionic starter app and provide a link to the repo? I'm also using Brave with Chromium 111, and the text is truncated correctly.

**Comment by matthew2564** [comment]
> Are you able to reproduce this issue in an Ionic starter app and provide a link to the repo? I'm also using Brave with Chromium 111, and the text is truncated correctly.

https://github.com/matthew2564/ion-select-testing

I can also see a similar affect when building to the iPad 9th gen iOS simulator, although this repo has a lot more in it which is why I started the above to check if there were other factors

Screenshot from sim
<img width="710" alt="Screenshot 2023-04-05 at 15 49 03" src="https://user-images.githubusercontent.com/33056264/230118396-3d241820-b2f5-4e57-925e-5d7511c423cc.png">

**Comment by liamdebeasi** [comment]
Thanks! I can reproduce this issue now. The problem is related to stacked labels in `ion-select`. We incorrectly assume that the selected text can take up 100% of the `ion-select` width when using stacked labels. However, this is not true because we need to account for the width of the select icon.

**Comment by matthew2564** [comment]
> Thanks! I can reproduce this issue now. The problem is related to stacked labels in `ion-select`. We incorrectly assume that the selected text can take up 100% of the `ion-select` width when using stacked labels. However, this is not true because we need to account for the width of the select icon.

Brilliant, and thank you! 

If you need any more info/examples etc, then just reach out.

Thanks!

**Comment by liamdebeasi** [comment]
Here is a dev build if you are interested in trying a proposed fix: `7.0.2-dev.11680709876.185fb94b`

**Comment by matthew2564** [comment]
Truncation is working fine when specifying `aria-label` and no `labelPlacement`
![Screenshot 2023-04-06 at 08 47 41](https://user-images.githubusercontent.com/33056264/230310285-bc5e6f85-828b-4d3f-9275-4ffdf6b029b8.png)

But the caret is not locked to the right hand side (when not using `labelPlacement`), it appears to be reactive to the length the value in the `ion-select`

![Screenshot 2023-04-06 at 08 48 38](https://user-images.githubusercontent.com/33056264/230311570-0e088f6c-812a-4f2e-8b93-28cc50468a51.png)

![Screenshot 2023-04-06 at 08 47 30](https://user-images.githubusercontent.com/33056264/230310486-7885af2b-c109-46fa-a714-eb7b43f8fc70.png)

Truncation does not work when specifying both `aria-label` and `labelPlacement="stacked"`

![Screenshot 2023-04-06 at 08 52 23](https://user-images.githubusercontent.com/33056264/230311114-304dba41-ee19-4b08-86be-617e7e454941.png)

**Comment by liamdebeasi** [comment]
> Truncation does not work when specifying both aria-label and labelPlacement="stacked"

I'm not able to reproduce this using the dev build and the following code:

```html
<ion-select
  placeholder="Select a fruit"
  aria-label="fruit"
  multiple="true"
  fill="outline"
  style="border: 3px solid blue;"
  labelPlacement="stacked"
  [value]="['apple','banana','orange']"
>
  <ion-select-option value="apple">Apple</ion-select-option>
  <ion-select-option value="banana">Banana</ion-select-option>
  <ion-select-option value="orange">Orange</ion-select-option>
</ion-select>
```

Just FYI your `package-lock.json` is still trying to install a nightly build of Ionic which does not have the proposed fix: https://github.com/matthew2564/ion-select-testing/blob/0093096c2bfc44a0f14f517c5f1009ae86d60b8c/package-lock.json#L2981

That could be why you are not seeing any improvements.

**Comment by matthew2564** [comment]
@liamdebeasi Ah yes nice spot, it now appears to be working my side using `7.0.2-dev.11680713209.14d9bc55`, thanks!

**Comment by matthew2564** [comment]
Hey @liamdebeasi, just confirming if the fix made it into 7.0.2? 

Tried to bump my boilerplate project and the fix seems to not be there...? https://github.com/matthew2564/ion-select-testing/pull/1
